aproxy.go

```diff
@@ -0,0 +1,404 @@
+//go:build linux
+
+package main
+
+import (
+	"bufio"
+	"encoding/binary"
+	"errors"
+	"flag"
+	"fmt"
+	"golang.org/x/crypto/cryptobyte"
+	"io"
+	"log"
+	"log/slog"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"syscall"
+	"unsafe"
+)
+
+var version = "0.2.0"
+
+type PrereadConn struct {
+	prereadStarted bool
+	prereadEnded   bool
+	prereadBuf     []byte
+	mu             sync.Mutex
+	conn           net.Conn
+}
+
+func (c *PrereadConn) StartPreread() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.prereadStarted {
+		panic("call StartPreread after preread has already started or ended")
+	}
+	c.prereadStarted = true
+}
+
+func (c *PrereadConn) RestorePreread() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if !c.prereadStarted || c.prereadEnded {
+		panic("call RestorePreread after preread has ended or hasn't started")
+	}
+	c.prereadEnded = true
+}
+
+func (c *PrereadConn) Read(p []byte) (n int, err error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.prereadEnded {
+		n = copy(p, c.prereadBuf)
+		bufLen := len(c.prereadBuf)
+		c.prereadBuf = c.prereadBuf[n:]
+		if n == len(p) || (bufLen > 0 && bufLen == n) {
+			return n, nil
+		}
+		rn, err := c.conn.Read(p[n:])
+		return rn + n, err
+	}
+	if c.prereadStarted {
+		n, err = c.conn.Read(p)
+		c.prereadBuf = append(c.prereadBuf, p[:n]...)
+		return n, err
+	}
+	return c.conn.Read(p)
+}
+
+func (c *PrereadConn) Write(p []byte) (n int, err error) {
+	return c.conn.Write(p)
+}
+
+func NewPrereadConn(conn net.Conn) *PrereadConn {
+	return &PrereadConn{conn: conn}
+}
+
+func PrereadSNI(conn *PrereadConn) (_ string, err error) {
+	conn.StartPreread()
+	defer conn.RestorePreread()
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("failed to preread TLS client hello: %w", err)
+		}
+	}()
+	typeVersionLen := make([]byte, 5)
+	n, err := conn.Read(typeVersionLen)
+	if n != 5 {
+		return "", errors.New("too short")
+	}
+	if err != nil {
+		return "", err
+	}
+	if typeVersionLen[0] != 22 {
+		return "", errors.New("not a TCP handshake")
+	}
+	msgLen := binary.BigEndian.Uint16(typeVersionLen[3:])
+	buf := make([]byte, msgLen+5)
+	n, err = conn.Read(buf[5:])
+	if n != int(msgLen) {
+		return "", errors.New("too short")
+	}
+	if err != nil {
+		return "", err
+	}
+	copy(buf[:5], typeVersionLen)
+	return extractSNI(buf)
+}
+
+func extractSNI(data []byte) (string, error) {
+	s := cryptobyte.String(data)
+	var version uint16
+	var random []byte
+	var sessionId []byte
+	var compressionMethods []byte
+	var cipherSuites []uint16
```

Comment:
You could use a `var ( ... )` block to avoid `var` on each line here. However, I'd say it's more common to define the variables closer to where they're used, so only define them before each block below.

Reply:
Gotcha, moved the variable definition.